<a href="https://colab.research.google.com/github/sachinprabhu007/ML-Expriments-/blob/master/MNIST_Keras_implementation_with_dropout_and_hidden_layers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
from keras.optimizers import RMSprop, Adam

np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
#OPTIMIZER = SGD() 
OPTIMIZER = Adam()

N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3

# data: shuffled and split between train and test sets
(X_train, y_train),(X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalize

X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

Using TensorFlow backend.


60000 train samples
10000 test samples


In [2]:
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))

model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))


model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_1 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_2 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1290      
__________

In [0]:
model.compile(loss='categorical_crossentropy',
             optimizer=OPTIMIZER,
             metrics=['accuracy'])

In [4]:
history = model.fit(X_train,Y_train,
                   batch_size=BATCH_SIZE,
                   epochs=NB_EPOCH,
                   verbose=VERBOSE,
                   validation_split = VALIDATION_SPLIT)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 3s 63us/step - loss: 0.5183 - acc: 0.8432 - val_loss: 0.1896 - val_acc: 0.9433
Epoch 2/20
48000/48000 [==============================] - 3s 56us/step - loss: 0.2348 - acc: 0.9297 - val_loss: 0.1407 - val_acc: 0.9577
Epoch 3/20
48000/48000 [==============================] - 3s 56us/step - loss: 0.1829 - acc: 0.9461 - val_loss: 0.1160 - val_acc: 0.9656
Epoch 4/20
48000/48000 [==============================] - 3s 56us/step - loss: 0.1555 - acc: 0.9534 - val_loss: 0.1093 - val_acc: 0.9683
Epoch 5/20
48000/48000 [==============================] - 3s 56us/step - loss: 0.1338 - acc: 0.9593 - val_loss: 0.1000 - val_acc: 0.9714
Epoch 6/20
48000/48000 [==============================] - 3s 55us/step - loss: 0.1223 - acc: 0.9628 - val_loss: 0.0896 - val_acc: 0.9737
Epoch 7/20
48000/48000 [==============================] - 3s 54us/step - loss: 0.1090 - acc: 0.9661 - val_loss: 0.0938 - val_acc

In [5]:
score = model.evaluate(X_test,Y_test, verbose = VERBOSE)
print('Test score:', score[0])
print('Test accuracy:', score[1])

10000/10000 [==============================] - 0s 28us/step
Test score: 0.07587413650284143
Test accuracy: 0.9786


**Layers** - 3 

**For epoch = 200**

10000/10000 [==============================] - 0s 28us/step

Test score: 0.07601802289183252

Test accuracy: 0.9764

**For epoch = 20**

10000/10000 [==============================] - 0s 26us/step

Test score: 0.18603398130089044

Test accuracy: 0.9462

**Only 2 layers  and epoch = 20** 

only 1 hidden layer


10000/10000 [==============================] - 0s 32us/step

Test score: 0.235391285315156

Test accuracy: 0.935


**3 layers** 

**DROPOUT = 0.3 **
epoch = 250

add dropout for 2 hidden layers

10000/10000 [==============================] - 0s 26us/step

Test score: 0.07751633399298881

Test accuracy: 0.9779

**epoch = 20**

10000/10000 [==============================] - 0s 27us/step

Test score: 0.1994345935150981

Test accuracy: 0.9399

**changing optimizers  , epoch = 20**

**with RMSProp()**

10000/10000 [==============================] - 0s 27us/step

Test score: 0.09262821478234255

Test accuracy: 0.9769

**with Adam()**

10000/10000 [==============================] - 0s 28us/step

Test score: 0.07587413650284143

Test accuracy: 0.9786




